extension/src/project/readOnlyDocument.ts

```diff
@@ -97,12 +97,35 @@ export class ReadonlyDocument implements vscode.TextDocument {
             this.lines = fileContent.split('\r\n');
             this.lineCount = this.lines.length;
         }
-        this.atomsForest = [];
-        if (languageId === 'autolisp'){
+    }
+
+    
+    // Converted this from a constant data feature into an on-demand feature that once used is essentially cached for future queries.
+    get atomsForest(): Array<string|Sexpression> {
+        if (this.languageId === 'autolisp') {
+            if (this._atomsForest && this._atomsForest.length > 0){
+                return this._atomsForest;
+            } else {
+                this.updateAtomsForest();
+                return this._atomsForest;
+            }
+        } else {
+            return [];
+        }
+    }
+    
+    // This was segregated from the atomsForest getter to support two primary use cases:
+    //      A force update that will be called on the workspace.onDidSaveTextDocument() saved event to keep the memory document in sync with the user input.
```

Comment:
Shall we consider this situation? If a file is opened, edited, and saved on VS Code, we should access it via. the ordinary TextDocument of VS.Code, not this ReadonlyDocument.
The use of this ReadonlyDocument is limited to read the current content from a file, and that's all. The reason why we have this class is because there're some VSCode APIs which are designed to work with a document, not with a string.